photosynthesis_metrics/perceptual.py

```diff
@@ -0,0 +1,280 @@
+                 weights: List[Union[float, torch.Tensor]] = [1.], replace_pooling: bool = False, distance: str = "mse",
+                 reduction: str = "mean", normalize_input: bool = True, normalize_features: bool = False) -> None:
+        r"""
+        Args:
+            feature_extractor: Name of model used to extract features. One of {`vgg16`, `vgg19`}
+            layers: List of string with layer names. Default: [`relu3_3`]
+            weights: List of float weight to balance different layers
+            replace_pooling: Flag to replace MaxPooling layer with AveragePooling. See [1] for details.
+            distance: Method to compute distance between features. One of {`mse`, `mae`}.
+            reduction: Reduction over samples in batch: "mean"|"sum"|"none"
+            normalize_input: If true, scales the input from range (0, 1) to the range the
+                pretrained VGG network expects, namely (-1, 1)
+            normalize_features: If true, unit-normalize each feature in channel dimension before scaling
+                and computing distance. See [2] for details.
+
+        References:
+            .. [1] Gatys, Leon and Ecker, Alexander and Bethge, Matthias
+            (2016). A Neural Algorithm of Artistic Style}
+            Association for Research in Vision and Ophthalmology (ARVO)
+            https://arxiv.org/abs/1508.06576
+    
+            .. [2] Zhang, Richard and Isola, Phillip and Efros, et al.
+            (2018) The Unreasonable Effectiveness of Deep Features as a Perceptual Metric
+            2018 IEEE/CVF Conference on Computer Vision and Pattern Recognition
+            https://arxiv.org/abs/1801.03924
+
+        """
+        super().__init__()
+
+        if feature_extractor == "vgg16":
+            self.model = vgg16(pretrained=True, progress=False)
+            self.layers = [VGG16_LAYERS[l] for l in layers]
+        elif feature_extractor == "vgg19":
+            self.model = vgg19(pretrained=True, progress=False)
+            self.layers = [VGG19_LAYERS[l] for l in layers]
+        else:
+            raise ValueError("Unknown feature extractor")
+
+        if replace_pooling:
+            for i, layer in enumerate(self.model.features):
+                if isinstance(layer, torch.nn.MaxPool2d):
+                    self.model.features[i] = torch.nn.AvgPool2d(kernel_size=2, stride=2, padding=0)
+
+        # Disable gradients
+        for param in self.model.parameters():
+            param.requires_grad_(False)
+
+        self.distance = {
+            "mse": nn.MSELoss(reduction='none'),
+            "mae": nn.L1Loss(reduction='none'),
+        }[distance]
+
+        self.weights = [torch.tensor(w) for w in weights]
+        mean = torch.tensor([0.485, 0.456, 0.406]) if normalize_input else torch.tensor([0., 0., 0.])
+        std = torch.tensor([0.229, 0.224, 0.225]) if normalize_input else torch.tensor([1., 1., 1.])
```

Comment:
Please move `torch.tensor([0.485, 0.456, 0.406])` and `torch.tensor([0.229, 0.224, 0.225])` to a separate variable with proper names which describe where these values are taken from